### ISSUE TITLE
AP alternate headwords

### ISSUE THREAD

**Issue by drdhaval2785** [sanskrit-lexicon/alternateheadwords#3]
Total 15 to examine.

```
1:atipAdanicf(vf)t:atipAdanivft:atipAdanicfvf
3:Agu(gU)r:AgUr:AgugU
3:ulla(lu)kasanam:ullukasanam:ullalusanam
5:u(U)hApohaH:UhApohaH:uUApohaH
1:Kawa(qa)kkikA:KaqakkikA:KawaqaikA
3:daSe(se)rakaH:daserakaH:daSesekaH
3:dagA(kA)rgalam:dakArgalam:dagAkAalam
1:damu(mU)nas:damUnas:damumUs
3:diDi(DI)zUH:diDIzUH:diDiDIH
1:parizka(ska)nda:pariskanda:parizkaska
1:pAMsa(Sa)na:pAMSana:pAMsaSa
1:picCa(cCi)kA:picCikA:picCacCi
3:piwaNkA(NgA)SaH:piwaNgASaH:piwaNkANgA
1:vArdDuzi(za):vArdDuza:vArdDuziza
3:vyaBi(BI)cAraH:vyaBIcAraH:vyaBiBIraH
```

**Comment by drdhaval2785** [comment]
Examined all.
All suggestions are correct.
Ready to be installed.
https://github.com/sanskrit-lexicon/alternateheadwords/commit/6baf1470ba4539e061b6d9c5128ba0103ab315e8

**Comment by gasyoun** [comment]
And not `AgUr:Agur` ?

**Comment by drdhaval2785** [comment]
@gasyoun You seems to have misread the entries.

code:presenthw:suggestedhw:scrap.

Last one has no relevance at all. Only the presenthw and suggestedhw has some relevance. The last member is a SCRAP.

**Comment by gasyoun** [comment]
`The last member is a SCRAP.` what use of them then?

**Comment by drdhaval2785** [comment]
> The last member is a SCRAP. what use of them then?

The bracket can be opened right or left. There are two possible resolutions for any given bracket.
They have to be generated and their relative weightages have to be checked to see which of these resolutions are more grammatically possible. So whatever had higher weightage has been put in the first place and the other one in the scrap category. 

But there are cases where the machine is not able to resolve the brackets with any confidence. 
They are marked with code '0'.
In such cases, both are of equal weightage. 

So, actual usage of this scrap data is when the code is '0'
or in rare circumstance,
Weightage assignment may have been wrong (after all it is probabilistic).

**Comment by gasyoun** [comment]
`relative weightages have to be checked to see which of these resolutions are more grammatically possible` - oh, so it's that advanced. All I did is write some rules in Excel, but none related to grammar.
`actual usage of this scrap data is when the code is '0'` - understood it now, thanks.

**Comment by drdhaval2785** [comment]
https://github.com/sanskrit-lexicon/alternateheadwords#correction-codes

This documents what are the logics used in deciding the code for correction.

Majority of them were covered.
I cross checked small sample in AP90.
Out of 40 I checked, 2 were not decided properly by machine, and they too were '0'. So, good enough decent output is there. False positives and false negatives are relatively few.

**Comment by gasyoun** [comment]
> https://github.com/sanskrit-lexicon/alternateheadwords#correction-codes

I've seen, thanks.

**Comment by funderburkjim** [comment]
This is the first one I've looked at.

As I understand it, the `data/AP/apahw3.txt` file is the end result, And I think I see how
to interpret it.

It seems that the ball is now in my court.

@drdhaval2785 Have you given thought how to integrate these alternate headwords into the
AP dictionary ?

I have one idea, but would like to hear ideas of others before pulling any programmatic triggers.

**Comment by funderburkjim** [comment]
Idea regarding format of apahw3 -

I see you are starting with ap90hw0.txt.

In order to use a line of apahw3.txt,  it might be useful to know what is the corresponding line within
ap90hw0.txt.  

Consider example. Here is original hw0 line

```
0043-1:atipAdanicf(vf)t:3642,3643
```

Your analyzed line is 

```
1:atipAdanicf(vf)t:atipAdanivft:atipAdanicfvf
```

I guess in this case I can match the original and analyzed via string `atipAdanicf(vf)t`.

This should work for this dictionary.

However, I am concerned that for some dictionaries, this technique (of matching 'key2' from hw0)
might not work.  However, matching on the beginning-ending line numbers (e.g. 3642,3643) should
always work. 

Suggestion for thought:  Modify program sequence to carry along the positional meta-data .
For example:

```
1:atipAdanicf(vf)t:atipAdanivft:atipAdanicfvf:3642,3643
```

For now, this change is not required.  But it might be useful for other dictionaries.

**Comment by gasyoun** [comment]
> In order to use a line of apahw3.txt, it might be useful to know what is the corresponding line within
> ap90hw0.txt.

Makes sense. But what if the line number changes? There have been such cases in the past. Or not in the original file, only in the manualchange files?

**Comment by drdhaval2785** [comment]
> @drdhaval2785 Have you given thought how to integrate these alternate
> headwords into the AP dictionary ?

I think (1) we should add them with some unique identifier at the end of
dictionary or (2) 1254.1 dot separated numbers in the given headword, just
like addenda pages in MW.
Second seems more feasible to me. We have precedence of MW to follow.

**Comment by drdhaval2785** [comment]
@gasyoun

> Makes sense. But what if the line number changes? There have been such cases in the past. Or not in the original file, only in the manualchange files?

In case of mismatch, Jim's pythons always spit out an error or log. Having
headwordvas primary matcher and lines as secondary help would only reduce
the mismatch cases. I see that I can easily preserve startline and endline
for Jim. Correcting code that way.

**Comment by gasyoun** [comment]
> I see that I can easily preserve startline and endline
> for Jim.

Oh, fine.

**Comment by drdhaval2785** [comment]
@funderburkjim Ready to install.

**Comment by drdhaval2785** [comment]
@funderburkjim
Ready to install.
